### ISSUE TITLE
Support a way for an attribute class's AllowMultiple value to be propagated to derived classes.

### ISSUE THREAD

**Issue by Korporal** [dotnet/runtime#49606]
## Background and Motivation

When creating attribute classes it is sometimes the case that we will create a base attribute class and then specialize that by creating several derived attribute classes. When we do this we can specify an attribute usage for the base class and set Inherited = true, this propagates that attribute usage to each derived class, but does not propagate the AllowMultiple property.

This means that we cannot prevent prevent different derived attribute from being applied together, despite the fact that the base attribute class has AllowMultiple = false. In cases where we want these derived attributes to always be mutually exclusive, additional runtime code must be included.

## Proposed API

Add a new property `AllowMultipleInheritors` to the AttributeUsage class:
```
    [AttributeUsage(AttributeTargets.Class, Inherited = true)]
    public sealed class AttributeUsageAttribute : Attribute
    {
        public AttributeUsageAttribute(AttributeTargets validOn);
        public bool AllowMultiple { get; set; }
        public bool Inherited { get; set; }
        public AttributeTargets ValidOn { get; }
	public bool AllowMultipleInheritors { get; set; }
    }
```
The effect being that the compiler will not allow multiple, different derived attributes to be applied to a code element.

## Usage Examples

```
[AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true, AllowMultipleInheritors = false]
public class AttributesBase : Attribute
{

}

public class DerivedAttributeA : AttributeBase
{

}

public class DerivedAttributeB : AttributeBase
{

}

public class Example
{
  [DerivedAttributeA()]
  [DerivedAttributeB()]
  public string Something { get; set; }
}

```

In this case, although the two attributes applied to the property `Something` are different, the compiler will not allow this and will generate an error message. The current implementation always allows multiple attributes like this to coexist even when they are derived from the same base class.

## Alternative Designs

At this point the only alternative means of prevent this is to incorporate runtime code that checks for the presence of multiple attributes that derive from the same base. Additionally for that code to be flexible we must include a way to indicate whether we want to apply this restriction or not, this can be done by (for example) including something like the `AllowMultipleInheritors` in our base class.

## Risks

It does not appear that this would be a breaking change if the default value of the new property `AllowMultipleInheritors` was true.

**Comment by jkotas** [comment]
> Alternative Designs

Roslyn analyzer?

**Comment by Korporal** [comment]
> > Alternative Designs
> 
> Roslyn analyzer?

I've never written an analyzer so can't say too much about this, but we would still some way to define when to apply the restriction and when not.

**Comment by jkotas** [comment]
Yes, the attributes to control it would have to come with the analyzer.

The advantage of the analyzers is that everybody can build them. It is not necessary to add them to the core platform to do what you need.

**Comment by jnm2** [comment]
> I've never written an analyzer so can't say too much about this, but we would still some way to define when to apply the restriction and when not.

You could put an attribute on the attribute to tell the analyzer what to do.

```cs
[DisallowMultipleInheritors]
[AttributeUsage(...)]
public abstract class SomeAttribute : Attribute
{
    // ...
}
```

**Comment by Korporal** [comment]
> > I've never written an analyzer so can't say too much about this, but we would still some way to define when to apply the restriction and when not.
> 
> You could put an attribute on the attribute to tell the analyzer what to do.
> 
> ```cs
> [DisallowMultipleInheritors]
> [AttributeUsage(...)]
> public abstract class SomeAttribute : Attribute
> {
>     // ...
> }
> ```

That's true, that would work, and is a good way to do this even if writing runtime code to check for this.

**Comment by buyaa-n** [comment]
> That's true, that would work, and is a good way to do this even if writing runtime code to check for this.

@Korporal is that means you are good now and decided to implement this with an analyzer?

**Comment by Korporal** [comment]
@buyaa-n - Not exactly, I've hardly spent any time on an analyzer and it would take me some time that I'm likely not going to get.

I did start looking into analyzers then found that the documentation for debugging them was unhelpful and so I posted suggestions and issues about that, that basically took up time and distracted me.

**Comment by canton7** [comment]
It would be nice to have an option to have this properly enforced by the compiler, so we can get compile-time guarantees that attributes are being used correctly (as `AttributeUsage` allows for other restriction types), rather than an analyzer which can be simply excluded by the user.

**Comment by buyaa-n** [comment]
Yeah, but that would be a compiler/roslyn feature i think

**Comment by canton7** [comment]
Indeed, but Roslyn would need `AttributeUsageAttribute` to let the user specify that attributes derived from a given base attribute should be mutually exclusive, so a runtime (BCL) change would be required first.

**Comment by joperezr** [comment]
[Triage]: While @canton7 point is correct that in order to get full support we would need to eventually add the attribute to the framework, I think the best path forward if we do want to add this feature to the framework would be to first make the proposal in roslyn to see if they would add support for this, and if so then we can come back and consider adding the attribute then, but it would be no use to add the attribute if there will be no roslyn support for it to begin with. Due to this, I'll go ahead and transfer this issue to that repo for now, but let me know if there is a problem with that.
